pkg/operator/target_config_reconciler.go

```diff
@@ -303,10 +326,282 @@ func (c TargetConfigReconciler) sync() error {
 	if err != nil {
 		return err
 	}
+	return nil
+}
+
+const kueueFinalizer = "kueue.openshift.io/finalizer"
 
+func (c *TargetConfigReconciler) addFinalizerToKueueInstance(kueue *kueuev1alpha1.Kueue) error {
+	if !controllerutil.ContainsFinalizer(kueue, kueueFinalizer) {
+		controllerutil.AddFinalizer(kueue, kueueFinalizer)
+		if _, err := c.operatorClient.Kueues(kueue.Namespace).Update(c.ctx, kueue, metav1.UpdateOptions{}); err != nil {
+			klog.ErrorS(err, "Failed to add finalizer to Kueue instance", "Kueue", kueue.Name)
+			return err
+		}
+		klog.Infof("Finalizer %s added to Kueue instance %s", kueueFinalizer, kueue.Name)
+	}
 	return nil
 }
 
+func (c *TargetConfigReconciler) removeFinalizerFromKueueInstance(kueue *kueuev1alpha1.Kueue) error {
+	if !controllerutil.ContainsFinalizer(kueue, kueueFinalizer) {
+		return nil
+	}
+
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		latest, getErr := c.operatorClient.Kueues(kueue.Namespace).Get(c.ctx, kueue.Name, metav1.GetOptions{})
+		if getErr != nil {
+			return getErr
+		}
+		controllerutil.RemoveFinalizer(latest, kueueFinalizer)
+		_, updateErr := c.operatorClient.Kueues(latest.Namespace).Update(c.ctx, latest, metav1.UpdateOptions{})
+		return updateErr
+	})
+
+	if err != nil {
+		klog.ErrorS(err, "Failed to remove finalizer from Kueue instance", "Kueue", kueue.Name)
+		return err
+	}
+	klog.Infof("Finalizer %s removed from Kueue instance %s", kueueFinalizer, kueue.Name)
+	// c.cleanUpOperatorRole()
+	return nil
+}
+
+func (c *TargetConfigReconciler) cleanUpResources() error {
+	ctx := context.TODO()
+	klog.Info("Pannaga deleted kueue cluster")
+	klog.Info("Starting cleanup of Kueue resources...")
+
+	var encounteredError error
```

Comment:
encounteredError has too many side effects in the function. Perhaps use  the [apimachinery](https://github.com/kubernetes/apimachinery/blob/v0.32.2/pkg/util/errors/errors.go)  package to aggregate a list of errors. You would use `[]error` to add an error to the list.